runtime/libc/wasi/wasi_impl_minimal.c

```diff
@@ -0,0 +1,1129 @@
+    }
+
+    ssize_t nread = 0;
+    while (nread < buf_len) {
+        size_t rc = read(urandom_fd, buf, buf_len);
+        if (rc < 0)
+            goto ERR_READ;
+        nread += rc;
+    }
+
+    close(urandom_fd);
+    return __WASI_ERRNO_SUCCESS;
+
+ERR_READ:
+    close(urandom_fd);
+    return wasi_fromerrno(errno);
+NO_UDEV:
+    do {
+        __wasi_size_t buf_cursor = 0;
+        for (; (buf_len - buf_cursor) >= 4; buf_cursor += 4) {
+            *((int*)(buf + buf_cursor)) = random();
+        }
+        for (; buf_cursor < buf_len; buf_cursor += 1) {
+            *(buf + buf_cursor) = random() % UINT8_MAX;
+        }
+    } while (0);
+
+    return __WASI_ERRNO_SUCCESS;
+}
+
+/**
+ * Temporarily yield execution of the calling thread similar to `sched_yield` in POSIX.
+ * This implementation ignores client calls and silently returns RC 0
+ *
+ * @return __WASI_ERRNO_SUCCESS
+ */
+__wasi_errno_t __wasi_sched_yield(void* wasi_context) {
+    return __WASI_ERRNO_SUCCESS;
+}
+
+/**
+ * Receive a message from a socket.
+ * Note: This is similar to `recv` in POSIX, though it also supports reading
+ * the data into multiple buffers in the manner of `readv`.
+ *
+ * @param fd
+ * @param ri_data_baseretptr List of scatter/gather vectors to which to store data.
+ * @param ri_data_len The length of the array pointed to by `ri_data`.
+ * @param ri_flags Message flags.
+ * @param ri_data_nbytes_retptr Number of bytes stored in ri_data flags.
+ * @param message_nbytes_retptr Number of bytes stored in message flags.
+ * @return status code
+ */
+__wasi_errno_t __wasi_sock_recv(void* wasi_context, __wasi_fd_t fd, const __wasi_iovec_t* ri_data, size_t ri_data_len,
+                                __wasi_riflags_t ri_flags, __wasi_size_t* ri_data_nbytes_retptr,
+                                __wasi_roflags_t* message_nbytes_retptr) {
+    return wasi_unsupported_syscall(__func__);
+}
+
+/**
+ * Send a message on a socket.
+ * Note: This is similar to `send` in POSIX, though it also supports writing
+ * the data from multiple buffers in the manner of `writev`.
+ *
+ * @param fd
+ * @param si_data_baseptr List of scatter/gather vectors to which to retrieve data
+ * @param si_data_len The length of the array pointed to by `si_data`.
+ * @param si_flags Message flags.
+ * @param nsent_retptr Number of bytes transmitted.
+ * @return status code
+ */
+__wasi_errno_t __wasi_sock_send(void* wasi_context, __wasi_fd_t fd, const __wasi_ciovec_t* si_data, size_t si_data_len,
+                                __wasi_siflags_t si_flags, __wasi_size_t* nsent_retptr) {
+    return wasi_unsupported_syscall(__func__);
+}
+
+/**
+ * Shut down socket send and receive channels.
+ *
+ * @param fd
+ * @param how Which channels on the socket to shut down.
+ * @return status code
+ */
+__wasi_errno_t __wasi_sock_shutdown(void* wasi_context, __wasi_fd_t fd, __wasi_sdflags_t how) {
+    /* similar to `shutdown` in POSIX. */
+    return wasi_unsupported_syscall(__func__);
+}
```

Comment:
It isn't clear to me why some functions are implemented, and others aren't. If there's a design/assumptions around this, it would be best to document those, perhaps in a comment at the head of the file?